File: java/com/countrygamer/arcanacraft/common/quom/Quom.java
package com.countrygamer.arcanacraft.common.quom;

import java.util.List;

import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Gui;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.MovingObjectPosition;
import net.minecraft.util.ResourceLocation;
import net.minecraft.world.World;
import net.minecraftforge.event.entity.player.PlayerInteractEvent;

import com.countrygamer.arcanacraft.common.ArcanaCraft;
import com.countrygamer.arcanacraft.common.extended.ExtendedArcanePlayer;
import com.countrygamer.countrygamercore.lib.UtilCursor;

public abstract class Quom {
	
	private String name;
	private String parentName;
	private int id;
	private ResourceLocation iconSource;
	
	public final int displayColumn, displayRow;
	
	public Quom(String name) {
		this(name, null);
	}
	
	public Quom(String name, Quom parent) {
		this(name, parent, 0, 0);
	}
	
	public Quom(String name, Quom parent, int col, int row) {
		this.name = name;
		this.parentName = parent == null ? "" : parent.getName();
		this.iconSource = new ResourceLocation(ArcanaCraft.pluginID, "textures/quoms/" + this.name
				+ ".png");
		
		this.displayColumn = col;
		this.displayRow = row;
		
		if (this.displayColumn < QuomRegistry.minDisplayColumn) {
			QuomRegistry.minDisplayColumn = this.displayColumn;
		}
		
		if (this.displayRow < QuomRegistry.minDisplayRow) {
			QuomRegistry.minDisplayRow = this.displayRow;
		}
		
		if (this.displayColumn > QuomRegistry.maxDisplayColumn) {
			QuomRegistry.maxDisplayColumn = this.displayColumn;
		}
		
		if (this.displayRow > QuomRegistry.maxDisplayRow) {
			QuomRegistry.maxDisplayRow = this.displayRow;
		}
		
		this.id = QuomRegistry.quomRegistry.size();
		QuomRegistry.registerQuom(this);
	}
	
	public String getName() {
		return this.name;
	}
	
	public int getID() {
		return this.id;
	}
	
	public void draw(Minecraft mc, Gui g, int x, int y) {
		mc.getTextureManager().bindTexture(this.iconSource);
		g.drawTexturedModalRect(x, y, 0, 0, 16, 16);
	}
	
	public void saveToNBT(NBTTagCompound tagCom) {
		tagCom.setString("name", this.name);
		tagCom.setString("parentName", this.parentName);
		tagCom.setInteger("id", this.id);
		tagCom.setString("iconSource_dir", this.iconSource.getResourceDomain());
		tagCom.setString("iconSource_path", this.iconSource.getResourcePath());
		
	}
	
	public void loadFromNBT(NBTTagCompound tagCom) {
		this.name = tagCom.getString("name");
		this.parentName = tagCom.getString("parentName");
		this.id = tagCom.getInteger("id");
		this.iconSource = new ResourceLocation(tagCom.getString("iconSource_dir"),
				tagCom.getString("iconSource_path"));
		
	}
	
	public Quom copy() {
		// NBTTagCompound tagCom = new NBTTagCompound();
		// this.saveToNBT(tagCom);
		// return Quom.getFromNBT(tagCom);
		return this;
	}
	
	@Override
	public boolean equals(Object obj) {
		if (obj instanceof Quom) {
			Quom that = (Quom) obj;
			return this.name.equals(that.name) && this.id == that.id
					&& this.parentName.equals(that.parentName);
		}
		return false;
	}
	
	public Quom getParent() {
		return QuomRegistry.getQuom(this.parentName);
	}
	
	public double getReachLength(Tiers.Cast castTier, boolean isCreative) {
		return 5.0D;
	}
	
	public boolean isSingleCast() {
		return true;
	}
	
	public void onUse_do(EntityPlayer player, ExtendedArcanePlayer arcanePlayer, World world,
			Tiers.Cast castTier, Tiers.MANUS manusTier) {
		// ArcanaCraft.logger.info("Quom");
		
		// arcanePlayer.decrementManus(Tiers.Cast.tierToInt(castTier) * 2);
		
		boolean usedOnEntity = false;
		MovingObjectPosition mop = UtilCursor.getMOPFromPlayer(world, player,
				this.getReachLength(castTier, player.capabilities.isCreativeMode));
		if (mop != null) {
			if (mop.typeOfHit == MovingObjectPosition.MovingObjectType.ENTITY) {
				// ArcanaCraft.logger.info("Found Entity");
				if (mop.entityHit instanceof EntityLivingBase) {
					usedOnEntity = this.onEntityUse(player, arcanePlayer, world,
							(EntityLivingBase) mop.entityHit, castTier);
				}
			}
			else {
				// ArcanaCraft.logger.info("Not An Entity");
			}
		}
		else {
			// ArcanaCraft.logger.info("Null MOP");
		}
		
		if (!usedOnEntity) {
			UtilCursor.MovingObjectPositionTarget mopT = UtilCursor.getBlockFromCursor(world,
					arcanePlayer.player,
					this.getReachLength(castTier, player.capabilities.isCreativeMode));
			if (mopT != null) {
				if (this.onUse(player, arcanePlayer, world, mopT.x, mopT.y, mopT.z, mopT.side,
						castTier)) return;
			}
			this.onRightClick(player, arcanePlayer, world, castTier);
		}
	}
	
	private boolean onEntityUse(EntityPlayer player, ExtendedArcanePlayer arcanePlayer,
			World world, EntityLivingBase entity, Tiers.Cast castTier) {
		return false;
	}
	
	public abstract boolean onUse(EntityPlayer player, ExtendedArcanePlayer arcanePlayer,
			World world, int x, int y, int z, int side, Tiers.Cast castTier);
	
	public abstract void onRightClick(EntityPlayer player, ExtendedArcanePlayer arcanePlayer,
			World world, Tiers.Cast castTier);
	
	public void checkForDiscovery(ExtendedArcanePlayer arcanePlayer, int type,
			PlayerInteractEvent.Action action, ItemStack itemStack) {
		if (this.canDiscover_base(arcanePlayer, type, action, itemStack)) {
			// ArcanaCraft.logger.info("Can discover " + this.getName());
			arcanePlayer.discoverQuom(this);
			return;
		}
		// ArcanaCraft.logger.info("Cannot discover " + this.getName());
	}
	
	public boolean canDiscover_base(ExtendedArcanePlayer arcanePlayer, int type,
			PlayerInteractEvent.Action action, ItemStack itemStack) {
		boolean hasLearnedParent = this.getParent() == null
				|| arcanePlayer.hasLearnedQuom(this.getParent());
		if (hasLearnedParent && !arcanePlayer.hasDiscoveredQuom(this)) {
			switch (type) {
				case 0: // item use
					return this.canDiscover_Use(arcanePlayer, action, itemStack);
				case 1: // craft
					return this.canDiscover_Craft(arcanePlayer, itemStack);
				case 2: // smelt
					return this.canDiscover_Smelt(arcanePlayer, itemStack);
			}
		}
		return false;
	}
	
	public boolean canDiscover_Use(ExtendedArcanePlayer arcanePlayer,
			PlayerInteractEvent.Action action, ItemStack itemStack) {
		return false;
	}
	
	public boolean canDiscover_Craft(ExtendedArcanePlayer arcanePlayer, ItemStack itemStack) {
		return false;
	}
	
	public boolean canDiscover_Smelt(ExtendedArcanePlayer arcanePlayer, ItemStack itemStack) {
		return false;
	}
	
	public boolean canLearnNormally(ExtendedArcanePlayer arcanePlayer) {
		return true;
	}
	
	public void addHoverInfo(ExtendedArcanePlayer player, List<String> info) {
	}
	
}

File: src/main/scala/com/countrygamer/arcanacraft/common/quom/Quom.scala
package com.countrygamer.arcanacraft.common.quom

import com.countrygamer.arcanacraft.common.ArcanaCraft
import com.countrygamer.arcanacraft.common.extended.ArcanePlayer
import com.countrygamer.arcanacraft.common.init.Quoms
import com.countrygamer.cgo.common.lib.util._
import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.{MovingObjectPosition, ResourceLocation}
import net.minecraft.world.World

/**
 *
 *
 * @author CountryGamer
 */
class Quom(private var name: String, parent: Quom, var col: Int, var row: Int) {

	private var parentName: String = ""
	private var id: Int = 0
	private var icon: ResourceLocation = null

	// Default Constructor
	{
		if (parent != null) {
			this.parentName = parent.getName()

		}

		this.icon = new ResourceLocation(ArcanaCraft.pluginID,
			"textures/quoms/" + this.name + ".png")

		if (this.col < Quoms.minDisplayColumn) {
			Quoms.minDisplayColumn = this.col
		}

		if (this.row < Quoms.minDisplayRow) {
			Quoms.minDisplayRow = this.row
		}

		if (this.col > Quoms.maxDisplayColumn) {
			Quoms.maxDisplayColumn = this.col
		}

		if (this.row > Quoms.maxDisplayRow) {
			Quoms.maxDisplayRow = this.row
		}

		this.id = Quoms.registerQuom(this)

	}

	// End Constructor

	// Other Constructors
	def this(name: String, col: Int, row: Int) {
		this(name, null, col, row)

	}

	def this(name: String, parent: Quom) {
		this(name, parent, 0, 0)

	}

	def this(name: String) {
		this(name, null)

	}

	// End Constructors

	def getName(): String = {
		this.name
	}

	def getID(): Int = {
		this.id
	}

	def getParent(): Quom = {
		Quoms.getQuom(this.parentName)
	}

	// Client Side things

	@SideOnly(Side.CLIENT)
	def draw(gui: net.minecraft.client.gui.Gui, x: Int, y: Int): Unit = {
		this.draw(gui, x, y, 0, 0, 0, 0)
	}

	@SideOnly(Side.CLIENT)
	def draw(gui: net.minecraft.client.gui.Gui, x: Int, y: Int, leftOffset: Int, rightOffset: Int,
			topOffset: Int, bottomOffset: Int): Unit = {
		UtilRender.bindResource(this.icon)
		UtilRender.drawTextureWithOffsets(
			gui, x, y, 0, 0, 16, 16, leftOffset, rightOffset, topOffset, bottomOffset)

	}

	// End Client Side things

	def saveToNBT(tagCom: NBTTagCompound): Unit = {
		tagCom.setString("name", this.name)
		tagCom.setString("parentName", this.parentName)
		tagCom.setInteger("id", this.id)
		tagCom.setInteger("col", this.col)
		tagCom.setInteger("row", this.row)
		tagCom.setString("iconSource_dir", this.icon.getResourceDomain())
		tagCom.setString("iconSource_path", this.icon.getResourcePath())

	}

	def loadFromNBT(tagCom: NBTTagCompound): Unit = {
		this.name = tagCom.getString("name")
		this.parentName = tagCom.getString("parentName")
		this.id = tagCom.getInteger("id")
		this.col = tagCom.getInteger("col")
		this.row = tagCom.getInteger("row")
		this.icon = new ResourceLocation(tagCom.getString("iconSource_dir"),
			tagCom.getString("iconSource_path"))

	}

	override def equals(obj: scala.Any): Boolean = {
		obj match {
			case that: Quom =>
				this.id == that.getID() && this.name.equals(that.getName()) &&
						this.parentName.equals(that.getParent().getName())
			case _ =>
				false
		}
	}

	//

	final def onUse_do(player: EntityPlayer, arcanePlayer: ArcanePlayer, world: World,
			tier: Int): Unit = {
		val reachLength: Double = this.getReachLength(tier, player.capabilities.isCreativeMode)

		var usedOnEnity: Boolean = false
		var usedOnBlock: Boolean = false
		var used: Boolean = false

		val mop: MovingObjectPosition = Cursor.rayTrace(player, reachLength)

		if (mop != null && mop.typeOfHit == MovingObjectPosition.MovingObjectType.ENTITY) {
			mop.entityHit match {
				case entity: EntityLivingBase =>
					usedOnEnity = this.onUseEntity(player, arcanePlayer, world,
						tier, mop.entityHit.asInstanceOf[EntityLivingBase])
				case _ =>
			}
		}

		if (!usedOnEnity) {
			val mopT: MovingObjectPositionTarget = Cursor.getBlockFromCursor(world, player, reachLength)
			if (mopT != null) {
				usedOnBlock = this
						.onUseBlock(player, arcanePlayer, world, mopT.x, mopT.y, mopT.z, mopT.side,
				            tier)
			}

			if (!usedOnBlock) {
				used = this.onUse(player, arcanePlayer, world, tier)
			}

		}

	}

	//

	def getReachLength(tier: Int, isCreative: Boolean): Double = {
		5.0D
	}

	def onUseEntity(player: EntityPlayer, arcanePlayer: ArcanePlayer, world: World,
			tier: Int, entity: EntityLivingBase): Boolean = {
		false
	}

	def onUseBlock(player: EntityPlayer, arcanePlayer: ArcanePlayer, world: World, x: Int, y: Int,
			z: Int, side: Int, tier: Int): Boolean = {
		false
	}

	def onUse(player: EntityPlayer, arcanePlayer: ArcanePlayer, world: World,
			tier: Int): Boolean = {
		false
	}

	//

	def addHoverInformation(arcanePlayer: ArcanePlayer, hoverInfo: java.util.List[String]): Unit = {
	}

}
